### PR TITLE
feat(user): support multiple invites

### PR DESCRIPTION
**Pull request by apoorvdixit88**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description
The endpoint `/invite_multiple`, gives support to invite more than one user at a time.

### Additional Changes

- [x] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## Motivation and Context
Currently we can only invite one user at a time.

## How did you test it?
Curl for invite_multiple:
```
curl --location 'http://localhost:8080/user/user/invite_multiple' \
--header 'Authorization: Bearer JWT' \
--data-raw '[
    {
    "email": "admin7@juspay.in",
    "name": "25",
    "role_id": "merchant_view_only"
},
{
    "email": "test2@juspay.in",
    "name": "25",  
    "role_id": "merchant_view_only"
},
{
    "email": "12345@juspay.n",
    "name": "25",
    "role_id": "merchant_view_only"
}

]'
```
Screenshot 2024-01-23 at 4 28 32 AM

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/api_models/src/user.rs [hunk 1]
```diff
@@ -89,6 +89,16 @@ pub struct InviteUserResponse {
...
 }
 
+#[derive(Debug, serde::Serialize)]
+pub struct InviteMultipleUserResponse {
+    pub email: pii::Email,
+    pub is_email_sent: bool,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub password: Option<Secret<String>>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub error: Option<String>,
```
Comment (ThisIsMani) [inline crates/api_models/src/user.rs:99]: Can't this be done by one flag over the struct?

File: crates/router/src/core/user.rs [hunk 2]
```diff
@@ -467,6 +473,180 @@ pub async fn invite_user(
     }
 }
 
+pub async fn invite_multiple_user(
+    state: AppState,
+    requests: Vec<user_api::InviteUserRequest>,
+    user_from_token: auth::UserFromToken,
```
Comment (ThisIsMani) [inline crates/router/src/core/user.rs:479]: These should be reversed.

File: crates/router/src/core/user.rs [hunk 3]
```diff
@@ -467,6 +473,180 @@ pub async fn invite_user(
...
+    } else if invitee_user
+        .as_ref()
+        .map_err(|e| e.current_context().is_db_not_found())
+        .err()
+        .unwrap_or(false)
+    {
+        handle_new_user_invitation(state, user_from_token, request).await
+    } else {
+        Err(UserErrors::InternalServerError.into())
+    }
```
Comment (ThisIsMani) [inline crates/router/src/core/user.rs:530]: Could this be moved to the core function? Let's give some responsibility to core function as well.

File: crates/router/src/core/user.rs [hunk 4]
```diff
@@ -467,6 +473,180 @@ pub async fn invite_user(
...
+}
+
+async fn handle_existing_user_invitation(
+    state: &AppState,
+    user_from_token: &auth::UserFromToken,
+    request: &user_api::InviteUserRequest,
+    invitee_user: storage_user::User,
```
Comment (ThisIsMani) [inline crates/router/src/core/user.rs:537]: You can directly take `UserFromStorage` here.

File: crates/router/src/utils/user.rs [hunk 5]
```diff
@@ -118,3 +118,15 @@ pub fn get_verification_days_left(
...
+    request: &user_api::InviteUserRequest,
+    error: Option<String>,
+) -> user_api::InviteMultipleUserResponse {
+    user_api::InviteMultipleUserResponse {
+        email: request.email.clone(),
+        is_email_sent: false,
+        password: None,
+        error,
+    }
+}
```
Comment (ThisIsMani) [inline crates/router/src/utils/user.rs:132]: We can inline this function, this is very specific to invite.

File: crates/router/src/routes/user.rs [hunk 6]
```diff
@@ -350,6 +350,23 @@ pub async fn invite_user(
     ))
     .await
 }
+pub async fn invite_multiple_user(
+    state: web::Data<AppState>,
+    req: HttpRequest,
+    payload: web::Json<Vec<user_api::InviteUserRequest>>,
+) -> HttpResponse {
+    let flow = Flow::InviteUser;
```
Comment (ThisIsMani) [inline crates/router/src/routes/user.rs:358]: Should we create a different flow for this?

Comment (apoorvdixit88) [reply crates/router/src/routes/user.rs:358]: Later we will be using this api only after deprecating the previous version. I guess this flow is fine.

Comment (racnan) [reply crates/router/src/routes/user.rs:358]: yes we should.

**Review by ThisIsMani — APPROVED** [review]

**Review by racnan — APPROVED** [review]

**Pull request merged**